contrib/babelfishpg_tsql/src/procedures.c

```diff
@@ -108,6 +112,29 @@ bool		sp_describe_first_result_set_inprogress = false;
 char	   *orig_proc_funcname = NULL;
 static bool is_supported_case_sp_describe_undeclared_parameters = true;
 
+/* xml handle hash table */
+static HTAB         *XMLHandleHashTable = NULL;
+static MemoryContext XMLHashtabContext = NULL;
+const  int           XML_HANDLE_START = 0;
+const  int           XML_HANDLE_INVALID = INT32_MAX/2; 
+const  int           MAX_XML_HANDLES_PER_SESSION = 4000;  /* Maximum xml handles per session (size of hash table)*/ 
+static int           current_xml_handle;
+int     	         get_next_xml_handle(void);
+void                 pltsql_create_xml_handle_htab(void);
+void                 pltsql_delete_xml_handle_entry(int  handle);
+int                  pltsql_insert_xml_handle_entry(xmltype *xml_data,xmltype *ns_data, int xml_data_length, int ns_data_length);
+
+typedef struct XMLHandleHashEnt           /* Entries of hash table */
+{
+	int         document_id;
+	int         namespace_id;
+	int         original_document_size_bytes;
+	int         original_namespace_document_size_bytes;
+	bool        is_namespace_null;
+	xmltype    *xml_data; 
+	xmltype    *ns_data; 
```

Comment:
What are the restrictions on xml_data & ns_data field acc. to ideal TSQL behaviour? Currently, there are no limitations for these fields, which can cause the bloat. Let's verify and add length restrictions if applicable.